File: src/tools/elevation/stitcher.ts
// import { equalPoints } from '../../index.js';

// import type { IsolineSegment, MarchingSquaresResult } from './marching_old.js';
// import type {
//   MValue,
//   Properties,
//   VectorLineString,
//   VectorMultiLineString,
//   VectorMultiLineStringFeature,
//   VectorPoint,
// } from '../../index.js';

// /** `${from.x}_${from.y}`: To Segment */
// export type SegmentLookup = Map<string, IsolineSegment[]>;

// export interface ElevationProperties extends Properties {
//   elev: number;
//   elevFt: number;
// }

// export type ElevationFeature = VectorMultiLineStringFeature<
//   Record<string, unknown>,
//   MValue,
//   ElevationProperties
// >;

// export function stitchMarchingSquareSegments(
//   marchingSquares: MarchingSquaresResult,
// ): ElevationFeature[] {
//   const result: ElevationFeature[] = [];

//   for (const [threshold, segments] of marchingSquares) {
//     // setup graph for threshold
//     const pointToSegments = new Map<string, IsolineSegment[]>();
//     for (const seg of segments) {
//       const p1 = `${seg.from.x}_${seg.from.y}`;
//       const p2 = `${seg.to.x}_${seg.to.y}`;
//       if (!pointToSegments.has(p1)) pointToSegments.set(p1, []);
//       if (!pointToSegments.has(p2)) pointToSegments.set(p2, []);
//       pointToSegments.get(p1)!.push(seg);
//       pointToSegments.get(p2)!.push(seg);
//     }

//     const coordinates: VectorMultiLineString = [];
//     // Stitching
//     for (const startSeg of segments) {
//       if (startSeg.visited) continue;

//       const line: VectorLineString = [startSeg.from, startSeg.to];
//       startSeg.visited = true;

//       // Grow forward from 'to'
//       growLine(line, startSeg.to, pointToSegments);
//       // Grow backward from 'from' (and reverse the result)
//       const backwardPoints: VectorLineString = [];
//       growLine(backwardPoints, startSeg.from, pointToSegments);
//       line.unshift(...backwardPoints.reverse());

//       if (line.length <= 2) continue;

//       // remap line to 0->1
//       for (const point of line) {
//         point.x = point.x / 32_768;
//         point.y = point.y / 32_768;
//       }
//       coordinates.push(line);
//     }

//     if (coordinates.length === 0) continue;
//     result.push({
//       type: 'VectorFeature',
//       properties: { elev: threshold, elevFt: threshold * 3.28084 },
//       geometry: { type: 'MultiLineString', coordinates, is3D: false },
//     });
//   }

//   return result;
// }

// function growLine(
//   line: VectorLineString,
//   currentPt: VectorPoint,
//   graph: Map<string, IsolineSegment[]>,
// ) {
//   while (true) {
//     const key = `${currentPt.x}_${currentPt.y}`;
//     const neighbors = graph.get(key) ?? [];
//     const nextSeg = neighbors.find((s) => !s.visited);
//     if (nextSeg === undefined) break;

//     nextSeg.visited = true;
//     // Figure out which end of the segment is the new point
//     const nextPt = equalPoints(nextSeg.from, currentPt) ? nextSeg.to : nextSeg.from;
//     line.push(nextPt);
//     currentPt = nextPt;
//   }
// }

File: rust/tools/elevation/stitcher.rs
use crate::tools::{GridPoint, IsolineSegment, MarchingSquaresResult, OrderedF64};
use alloc::{collections::BTreeMap, rc::Rc};
use core::cell::RefCell;
use s2json::{MValue, VectorFeature, VectorGeometry, VectorLineString, VectorMultiLineString};
use serde::{Deserialize, Serialize};

/// GridPoint: To Segment
pub type SegmentLookup = BTreeMap<GridPoint, Vec<Rc<RefCell<IsolineSegment>>>>;

/// Contour Properties
#[derive(Debug, Default, Clone, Copy, PartialEq, Serialize, Deserialize, MValue)]
pub struct ContourProperties {
    /// Elevation in meters
    pub elev: f64,
    /// Elevation in feet
    #[serde(rename = "elevFt")]
    pub elev_ft: f64,
}

/// Stitch all segments together
pub fn stitch_marching_square_segments(
    marching_squares: MarchingSquaresResult,
) -> Vec<VectorFeature> {
    let mut result: Vec<VectorFeature> = vec![];

    for (OrderedF64(threshold), mut segments) in marching_squares {
        // setup graph for threshold
        let mut point_to_segments: SegmentLookup = BTreeMap::new();
        for seg in &segments {
            let IsolineSegment { from: p1, to: p2, .. } = *seg.borrow();
            let p1_set = point_to_segments.entry(p1).or_insert(vec![]);
            p1_set.push(seg.clone());
            let p2_set = point_to_segments.entry(p2).or_insert(vec![]);
            p2_set.push(seg.clone());
        }

        let mut coordinates: VectorMultiLineString = vec![];
        // Stitching
        for start_seg in &mut segments {
            if start_seg.borrow().visited {
                continue;
            }

            let mut forward_points: VectorLineString =
                vec![start_seg.borrow().from.into(), start_seg.borrow().to.into()];
            start_seg.borrow_mut().visited = true;

            // Grow forward from 'to'
            grow_line(&mut forward_points, start_seg.borrow().to, &mut point_to_segments);
            // Grow backward from 'from' (and reverse the result)
            let mut backward_points: VectorLineString = vec![];
            grow_line(&mut backward_points, start_seg.borrow().from, &mut point_to_segments);
            // merge forward and backward
            let mut line: VectorLineString = vec![];
            line.extend(backward_points.iter().rev().cloned());
            line.extend(forward_points.drain(..));

            if line.len() <= 2 {
                continue;
            }
            coordinates.push(line);
        }

        if coordinates.is_empty() {
            continue;
        }
        result.push(VectorFeature::new_wm(
            None,
            (ContourProperties { elev: threshold, elev_ft: threshold * 3.28084 }).into(),
            VectorGeometry::new_multilinestring(coordinates, None),
            None,
        ));
    }

    result
}

fn grow_line(line: &mut VectorLineString, mut curr_pt: GridPoint, graph: &mut SegmentLookup) {
    loop {
        let Some(neighbors) = graph.get_mut(&curr_pt) else {
            break;
        };
        let Some(next_seg) = neighbors.iter_mut().find(|s| !s.borrow().visited) else {
            break;
        };

        let next_seg = &mut next_seg.borrow_mut();

        next_seg.visited = true;
        // Figure out which end of the segment is the new point
        let next_pt = if next_seg.from == curr_pt { next_seg.to } else { next_seg.from };
        line.push(next_pt.into());
        curr_pt = next_pt;
    }
}
